File: ProductionConfigTool/SerialHandler.cs
using System;
using System.IO.Ports;
using System.Threading;

namespace ProductionTesterApp
{
    public class SerialHandler
    {
        private SerialPort serialPort;
        private bool isConnected = false;

        /// <summary>
        /// 시리얼 연결 핸들러 초기화
        /// </summary>
        /// <param name="portName">COM 포트 이름 (예: "COM3")</param>
        /// <param name="baudRate">통신 속도 (기본값: 115200)</param>
        /// <param name="timeout">읽기 타임아웃 (초 단위)</param>
        public SerialHandler(string portName, int baudRate = 115200, int timeout = 1)
        {
            try
            {
                serialPort = new SerialPort
                {
                    PortName = portName,
                    BaudRate = baudRate,
                    DataBits = 8,
                    Parity = Parity.None,
                    StopBits = StopBits.One,
                    ReadTimeout = timeout * 1000,
                    WriteTimeout = timeout * 1000
                };

                serialPort.Open();
                isConnected = true;
            }
            catch (Exception ex)
            {
                throw new Exception($"시리얼 포트 연결 실패: {ex.Message}");
            }
        }

        /// <summary>
        /// 데이터 패킷을 시리얼 연결을 통해 전송
        /// </summary>
        /// <param name="data">전송할 데이터 패킷</param>
        /// <returns>성공 여부</returns>
        public bool SendPacket(byte[] data)
        {
            if (!isConnected || serialPort == null)
                return false;

            try
            {
                serialPort.Write(data, 0, data.Length);
                return true;
            }
            catch (Exception)
            {
                return false;
            }
        }

        /// <summary>
        /// 시리얼 연결에서 응답 읽기
        /// </summary>
        /// <param name="expectedBytes">예상되는 바이트 수</param>
        /// <param name="timeout">타임아웃 (초 단위)</param>
        /// <returns>수신된 데이터</returns>
        public byte[] ReadResponse(int expectedBytes = 40, int timeout = 5)
        {
            if (!isConnected || serialPort == null)
                return null;

            try
            {
                // 기존 타임아웃 설정 임시 저장
                int originalTimeout = serialPort.ReadTimeout;
                serialPort.ReadTimeout = timeout * 1000;

                byte[] buffer = new byte[expectedBytes];
                int bytesRead = 0;
                DateTime startTime = DateTime.Now;

                // 예상 바이트 수만큼 읽거나 타임아웃까지 대기
                while (bytesRead < expectedBytes)
                {
                    if ((DateTime.Now - startTime).TotalSeconds > timeout)
                        break;

                    if (serialPort.BytesToRead > 0)
                    {
                        int count = serialPort.Read(buffer, bytesRead, Math.Min(serialPort.BytesToRead, expectedBytes - bytesRead));
                        bytesRead += count;
                    }
                    else
                    {
                        Thread.Sleep(10);
                    }
                }

                // 타임아웃 설정 복원
                serialPort.ReadTimeout = originalTimeout;

                if (bytesRead == 0)
                    return null;

                // 실제로 읽은 바이트 수에 맞게 배열 크기 조정
                if (bytesRead != expectedBytes)
                {
                    byte[] result = new byte[bytesRead];
                    Array.Copy(buffer, result, bytesRead);
                    return result;
                }

                return buffer;
            }
            catch (Exception)
            {
                return null;
            }
        }

        /// <summary>
        /// 명령 전송 및 응답 대기
        /// </summary>
        /// <param name="commandCode">명령 코드</param>
        /// <param name="data">명령과 함께 전송할 데이터 (선택사항)</param>
        /// <param name="waitForResponse">응답 대기 여부</param>
        /// <returns>응답 데이터 (waitForResponse가 true인 경우), 그렇지 않은 경우 null</returns>
        public byte[] SendCommand(byte commandCode, byte[] data = null, bool waitForResponse = true)
        {
            if (!isConnected || serialPort == null)
                return null;

            try
            {
                // 명령 패킷 구성
                byte[] commandPacket;
                if (data != null && data.Length > 0)
                {
                    commandPacket = new byte[data.Length + 1];
                    commandPacket[0] = commandCode;
                    Array.Copy(data, 0, commandPacket, 1, data.Length);
                }
                else
                {
                    commandPacket = new byte[] { commandCode };
                }

                // 명령 전송
                serialPort.Write(commandPacket, 0, commandPacket.Length);

                // 응답 대기
                if (waitForResponse)
                {
                    return ReadResponse();
                }

                return null;
            }
            catch (Exception)
            {
                return null;
            }
        }

        /// <summary>
        /// 장치 응답성 확인
        /// </summary>
        /// <returns>장치가 응답하는 경우 true, 그렇지 않은 경우 false</returns>
        public bool CheckDeviceStatus()
        {
            if (!isConnected || serialPort == null)
                return false;

            try
            {
                // 상태 확인 명령 전송 (0x01)
                byte[] response = SendCommand(0x01);
                return response != null && response.Length > 0;
            }
            catch (Exception)
            {
                return false;
            }
        }

        /// <summary>
        /// 시리얼 연결 닫기
        /// </summary>
        public void Close()
        {
            if (serialPort != null && serialPort.IsOpen)
            {
                serialPort.Close();
                isConnected = false;
            }
        }
    }
}

File: ProductionConfigTool/serial_handler.py
import serial
import time
import streamlit as st

class SerialHandler:
    """
    Handler for serial communication with production line devices
    """
    
    def __init__(self, port, baudrate=115200, timeout=1):
        """
        Initialize the serial connection
        
        Args:
            port (str): COM port to connect to
            baudrate (int): Baud rate (default 115200)
            timeout (int): Read timeout in seconds
        """
        try:
            self.serial = serial.Serial(
                port=port,
                baudrate=baudrate,
                bytesize=serial.EIGHTBITS,
                parity=serial.PARITY_NONE,
                stopbits=serial.STOPBITS_ONE,
                timeout=timeout
            )
            
            # Wait for connection to establish
            time.sleep(0.5)
            
        except Exception as e:
            raise Exception(f"Serial port connection failed: {str(e)}")
    
    def send_packet(self, data):
        """
        Send a data packet over the serial connection
        
        Args:
            data (bytes): Data packet to send
        
        Returns:
            bool: True if successful, False otherwise
        """
        if not self.serial.is_open:
            raise Exception("Serial port is not open")
        
        try:
            bytes_written = self.serial.write(data)
            return bytes_written == len(data)
        except Exception as e:
            raise Exception(f"Failed to send data: {str(e)}")
    
    def read_response(self, expected_bytes=40, timeout=5):
        """
        Read a response from the serial connection
        
        Args:
            expected_bytes (int): Number of bytes to read
            timeout (int): Timeout in seconds
            
        Returns:
            bytes: Received data
        """
        if not self.serial.is_open:
            raise Exception("Serial port is not open")
        
        start_time = time.time()
        buffer = bytearray()
        
        # Wait until we receive the expected number of bytes or timeout
        while (len(buffer) < expected_bytes) and (time.time() - start_time < timeout):
            if self.serial.in_waiting > 0:
                buffer += self.serial.read(self.serial.in_waiting)
            time.sleep(0.01)
        
        if len(buffer) < expected_bytes:
            raise Exception(f"Timeout waiting for response. Received {len(buffer)}/{expected_bytes} bytes")
        
        return bytes(buffer)
    
    def send_command(self, command_code, data=None, wait_for_response=True):
        """
        Send a command and wait for response
        
        Args:
            command_code (int): Command code
            data (bytes): Optional data to send with command
            wait_for_response (bool): Whether to wait for a response
            
        Returns:
            bytes: Response data if wait_for_response is True, else None
        """
        # Create command packet
        packet = bytearray([0xDA, command_code])  # STX and command code
        
        if data:
            packet.extend(data)
        
        # Add packet length
        data_length = len(packet) - 2  # Subtract STX and command code
        packet.insert(2, data_length)
        
        # Add ETX
        packet.append(0x25)
        
        # Send the packet
        self.send_packet(packet)
        
        # Wait for response if required
        if wait_for_response:
            return self.read_response()
        
        return None
    
    def check_device_status(self):
        """
        Check if the device is responsive
        
        Returns:
            bool: True if device is responsive, False otherwise
        """
        try:
            # Send a simple status check command
            response = self.send_command(0x01)  # Assuming 0x01 is status check command
            
            # Verify the response has correct format
            if response and len(response) >= 3:
                if response[0] == 0xDA and response[-1] == 0x25:
                    return True
            
            return False
        except Exception:
            return False
    
    def close(self):
        """Close the serial connection"""
        if hasattr(self, 'serial') and self.serial.is_open:
            self.serial.close()
